### PR TITLE
Add support for detailed profiling with native TF2 training

### PR DESCRIPTION
**Pull request by ndodda-amazon**
### Description of changes:
Followup to #500 to add detailed profiling support for native TF2 training.

#### Style and formatting:

I have run `pre-commit install` to ensure that auto-formatting happens with every commit.

#### Issue number, if available

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: smdebug/tensorflow/keras.py [hunk 1]
```diff
@@ -730,6 +730,56 @@ def _prepare_collections_for_tf2(self):
...
         self.prepared_tf2_collections = True
 
+    def _stop_detailed_profiling(self, current_step=None):
+        """Stop detailed profiling if the TF profiler is supported in the current TF version and detailed profiling
+        was already started.
+        """
+        if not is_profiler_supported_for_tf_version() or not self.is_detailed_profiling:
+            return
+
+        if current_step is None:
```
Comment (NihalHarish) [inline smdebug/tensorflow/keras.py:740]: Remove this `if-check` if `self.mode_steps[self.mode]` has `current_step` already.

File: smdebug/tensorflow/keras.py [hunk 2]
```diff
@@ -730,6 +730,56 @@ def _prepare_collections_for_tf2(self):
...
+            current_step = self.mode_steps[self.mode]
+
+        self.logger.info(f"Disabling TF profiler on step: ={current_step}")
+        stop_tf_profiler(
+            tf_profiler=self.tf_profiler,
+            log_dir=self._log_dir,
+            start_time_us=self.tf_profiler_start_time_in_micros,
+        )
+        self.is_detailed_profiling = False
```
Comment (NihalHarish) [inline smdebug/tensorflow/keras.py:749]: nit: write a similar function for `start_detailed_profiling`

File: smdebug/tensorflow/keras.py [hunk 3]
```diff
@@ -730,6 +730,56 @@ def _prepare_collections_for_tf2(self):
...
+
+    def _handle_detailed_profiling(self, current_step):
```
Comment (NihalHarish) [inline smdebug/tensorflow/keras.py:751]: nit: try to come up with a better name `_handle_detailed_profiling`

File: smdebug/tensorflow/keras.py [hunk 4]
```diff
@@ -730,6 +730,56 @@ def _prepare_collections_for_tf2(self):
...
+        if not is_profiler_supported_for_tf_version():
+            return
+
+        if self.profiler_config_parser.should_save_metrics(
+            MetricsCategory.DETAILED_PROFILING, current_step
+        ):
+            if not self.is_detailed_profiling:
+                self._log_dir = TraceFileLocation.get_detailed_profiling_log_dir(
+                    self.profiler_config_parser.config.local_path,
+                    "tensorflow",
```
Comment (NihalHarish) [inline smdebug/tensorflow/keras.py:769]: See if we can replace this with a variable

File: tests/profiler/tensorflow2/test_native_tf2_profiler.py [hunk 5]
```diff
@@ -108,6 +111,33 @@ def _verify_tensor_names(out_dir):
...
+def _verify_detailed_profiling(out_dir):
+    """
+    This verifies the number of events when detailed profiling is enabled.
+    """
+    t_events = TensorboardProfilerEvents()
+
+    # get tensorboard timeline files
+    files = []
+
+    for path in Path(os.path.join(out_dir, "framework")).rglob(f"*{TENSORBOARDTIMELINE_SUFFIX}"):
```
Comment (NihalHarish) [inline tests/profiler/tensorflow2/test_native_tf2_profiler.py:123]: Try to make this a helper function or a `fixture` if possible

**Review by NihalHarish — CHANGES_REQUESTED** [review]

File: smdebug/tensorflow/keras.py (hunk 1 above)
Comment (ndodda-amazon) [reply smdebug/tensorflow/keras.py]: Done

File: smdebug/tensorflow/keras.py (hunk 2 above)
Comment (ndodda-amazon) [reply smdebug/tensorflow/keras.py:767]: Done

File: smdebug/tensorflow/keras.py (hunk 3 above)
Comment (ndodda-amazon) [reply smdebug/tensorflow/keras.py]: Used `start_or_stop_detailed_profiling`

File: smdebug/tensorflow/keras.py (hunk 4 above)
Comment (ndodda-amazon) [reply smdebug/tensorflow/keras.py]: Done, used the existing `FRAMEWORK` enum

File: tests/profiler/tensorflow2/test_native_tf2_profiler.py (hunk 5 above)
Comment (ndodda-amazon) [reply tests/profiler/tensorflow2/test_native_tf2_profiler.py]: Made it a helper function.

**Review by NihalHarish — APPROVED** [review]

**Pull request merged**